throttle, UIButton.setBackgroundColor, PhotoEditCoordinator 관련 리팩토링

Hansolkkim opened the pull request:
#### 1. throttle 관련
- 이전에 말씀드렸던 `dimmedView`의 2중 터치가 가능하여 `viewController.dismiss(_:)`가 2번 호출되는 문제가 있어,
  2중 터치가 웬만하면 필요하지 않다고 판단해 대부분의 버튼과 dimmedView의 rx binding 시에 throttle을 거치도록 했습니다.
  약 0.4초 정도 동안의 터치를 1번의 터치 이벤트로 받아 들이도록 했고, 실제 테스트해보니까 버벅임이 느껴진다거나 하는 느낌은 없었습니다!

#### 2. UIButton.setBackgroundColor 관련
- #97 이슈에서 말씀드린 것처럼, 불필요하게 sRGB 색영역을 사용하지 않고 색상이 많이 필요하지 않은 경우 적절한(더 낮은 복잡도를 가지는) 색 영역을 선택하도록 하기 위해 `UIButton.setBackgroundColor(_:_:)` 메소드에서 `UIGraphicsImageRenderer`를 사용해 버튼의 배경색을 그리도록 했습니다.

#### 3. PhotoEditCoordinator 관련
- 이전 회의 때 말씀드린거 처럼, PhotoEditVC가 재활용될 때 ColorSelectVC, StyleSelectVC가 모두 함께 재활용될 필요가 있는데 이렇게 되면 하나의 Coordinator 객체가 가지는 책임이 불필요하게 커지는 문제가 생긴다고 판단하여 PhotoEditVC, ColorSelectVC, StyleSelectVC를 관리하는 `PhotoEditCoordinator`를 생성하여 AlbumCoordinator, GoalCoordinator에서 PhotoEditVC, ColorSelectVC, StyleSelectVC를 관리하는 책임을 조금 덜어내었습니다.
- 또한 PhotoEditView의 Mode를 나누었습니다.
  기존에 AlbumView에서 사용하던 PhotoEditView는 사진의 변경이 없는데 "저장" 버튼이 활성화될 경우 Realm과의 불필요한 통신이 생긴다고 판단하여 "저장"버튼의 활성화를 사진 데이터의 변경에 의해 결정되도록 했습니다.
  하지만 GoalListView에서 사용하는 PhotoEditView에서의 사진은 생성 당시에 입력된 정보와 초깃값을 가지는 데이터로 곧바로 저장될 수 있도록 할 필요가 있었기 때문에 "저장" 버튼의 활성화를 따로 관리할 필요가 없다고 생각했습니다.
  따라서 제드의 추천대로 **PhotoEditView의 모드를 2개로 나누어** AlbumView, GoalListView에서 각각 모드를 설정해 "저장" 버튼의 활성화에 차이를 두도록 했습니다.

Hansolkkim committed:
refactor: [#96] 버튼, DimmedView에 throttle 오퍼레이터 기능 추가

- 기존에 버튼, DimmedView들 중 연속 터치가 가능하여 의도치 않은 다중 실행이 되는 경우가 있어 모든 버튼, DimmedView에 RxSwift의 throttle 오퍼레이터를 사용하여 연속 터치를 제한함.

Hansolkkim committed:
refactor: [#97] setBackgroundColor 메소드 리팩토링

- UIButton을 extension해서 만든 `setBackgroundColor(_:UIColor? for:UIControl.state)` 메소드에서 `UIGraphicsBeginImageContext`를 사용해 배경색을 가지는 이미지를 렌더링하고 있었는데, `UIGraphicsBeginImageContext`를 사용해 이미지를 렌더링한다면 sRGB 색영역을 강제적으로 사용하여 메모리 낭비가 있어, 색영역을 컴파일러가 선택할 수 있도록 `UIGraphicsImageRenderer`를 사용해 이미지를 렌더링하는 방식으로 메소드 리팩토링 진행

Hansolkkim committed:
chore: Coordinator의 메소드 정리

- AlbumCoordinator에서 네비게이션을 위한 메소드들이 다 같은 형태인데 여러 메소드로 나뉘어져 있어, 이를 하나로 합침.
- GoalCoordinator에서 잘못 지어진 메소드의 이름을 수정하고, PhotoEditView를 pop하는 로직을 수정.

Hansolkkim committed:
refactor: [#99] PhotoEditCoordinator 분리

- ParentCoordinator들에서 공통적으로 사용할 ChildCoordinator를 제거하는 로직을 Coordinator protocol extension으로 정의
- AlbumCoordinator로부터 PhotoEditCoordinator 분리

Hansolkkim committed:
refactor: [#99] GoalCoordinator의 child로 PhotoEditCoordinator 사용하도록 수정

- 사진을 찍은 후 해당 사진의 속성을 수정해주는 기능을 추가하면서, PhotoEditVC, ColorSelectVC, StyleSelectVC를 모두 GoalCoordinator에서도 관리해주도록 하면서 GoalCoordinator의 역할이 너무 많아져, 해당 기능을 PhotoEditCoordinator가 책임지도록 수정

Hansolkkim committed:
refactor: [#99] PhotoEditView에 모드 추가

- PhotoEditCoordinator가 새로운 사진 등록시/ 기존 사진의 속성 변경 시 2번 사용되는데, 두 경우에 따라 다르게 표시되어야 할 "저장"버튼의 상태를 결정해주기 위해 PhotoEditMode 추가

Hansolkkim commented on rabit/rabit/Album/AlbumCoordinator.swift:
```diff
@@ -57,82 +36,27 @@ final class AlbumCoordinator: Coordinator, PhotoEditNavigation, AlbumNavigation,
+            photoStream: photoStream
         )
+        photoEditCoordinator.parentCoordiantor = self
+        photoEditCoordinator.didChangePhoto
+            .bind(to: didChangePhoto)
+            .disposed(by: disposeBag)
```
PhotoEditView에서 "저장" 버튼이 눌렸음을 PhotoEditCoordinator -> AlbumCoordinator로 알려주는 방법이 필요했습니다.

AlbumCoordinator의 AlbumNavigation에 해당하는 `didChangePhoto` 프로퍼티는 AlbumCoordinator가 사진이 변경되어 "저장" 버튼이 눌렸을 경우 트리거되는 Observer 프로퍼티인데, 이를 PhotoEditCoordinator에 넘겨주는게 조금 어색하다고 생각되었습니다. AlbumNavigation과 관련있는 프로퍼티이기 때문에 PhotoEditCoordinator가 받는게 어색해보여 위처럼 PhotoEditCoordinator를 생성하는 당시에 바인딩해주도록 했습니다..
근데 지금 다시 보니 여기서 이걸 처리하는거도 그리 깔끔해보이지는 않네요..ㅠ

junu0516 replied:
이건 코드를 좀 돌려보고.. 어떤 느낌인지 같이 볼께요!

Hansolkkim replied:
코드 설명을 첨부하자면, `didChangePhoto` 프로퍼티는 사진 속성 수정 화면인 PhotoEditView에서 "사진의 속성이 변경되었는지"에 따라 AlbumView에서 AlbumCollectionView의 reload를 할 것인지 안할 것인지를 결정할 때 사용하려고 만들어놓았던 프로퍼티입니다.
그런데 AlbumView와 PhotoEditView가 다른 코디네이터에 속하도록 수정이 되어, PhotoEditView에서 사진 속성 변경 되었음을 AlbumView에 알리는 방법이 애매해졌어요..ㅋㅋ

```
PhotoEditView의 저장 버튼을 누름 
-> 사진 정보를 Realm에 업데이트 
-> 업데이트가 정상적으로 되면 (AlbumViewModel이 바라보고있는, AlbumCoordinator의 프로퍼티인) didChangePhoto에 신호 보냄
 -> AlbumView가 신호를 받으면 AlbumCollectionView의 reload를 유도함
```

이런 흐름이었는데, Coordinator 분리로 아래의 흐름처럼 변경하려고 하는데 방법이 조금 어색해보여요 지금은 ..ㅎㅎ

```
PhotoEditView의 저장 버튼을 누름
-> PhotoEditView의 저장 버튼을 누름 
-> 사진 정보를 Realm에 업데이트 
-> 업데이트가 정상적으로 되면(PhotoEditCoordinator의 프로퍼티인) didChangePhoto에 신호 보냄 
-> 이로 인해 AlbumCoordinator의 프로퍼티인 didChangePhoto 또한 신호를 받게 됨 
-> AlbumView가 AlbumCollectionView의 reload를 유도함
```

junu0516 replied:
> @junu0516 리뷰 반영 바로 했습니다ㅎㅎ
> 
> AlbumCoordinator와 PhotoEditCoordinator 관련해서는 고민을 좀 해봤는데, Delegate 패턴을 이용해서 `PhotoEditCoordinatorDelegate`와 같은 프로토콜을 선언하고 AlbumCoordinator가 이를 채택하도록 하여 하위 코디네이터인 PhotoEditCoordinator가 특정 상황에 상위 코디네이터인 AlbumCoordinator: PhotoEditCoordinatorDelegate의 특정 메소드를 호출하도록 하든지, 아니면 이전처럼 특정 프로퍼티를 호출하는 것도 방법이 되겠다 생각이 들었습니다ㅎㅎ

@Hansolkkim 
요건 혹시 하위가 상위에 동작 끝난거 알릴 때 Rx대신 Delegate 패턴으로 대체하는 걸로 이해하면 될까요!!?
Delegate 패턴 쓰면 ㅋㅋ 제안하셨던 네이밍이 훨씬 자연스러울 것 같아요!!

하나 궁금한게 보통 RxSwift 사용하는 프로젝트에서 Rx 바인딩이랑 Delegate 패턴 중에 어떤 식으로 객체 간에 서로 소통하는 방식이 보편적인가 하는 생각이 드네요ㅎㅎㅎ
(이전에 이슈트래커 할 때 Deleagte -> Rx로 수정했던 기억이 있는데 지금 상황은 제가 맞게 이해한거면 그 때랑 반대길래 보통 현업에서 어떻게 할까 생각들었습니다 ㅎㅎ)

junu0516 replied:
ㅠ 객체가 하나 더 늘어나다보니 한단계 더 깊어지는건 불가피해보이네요 지금은

Hansolkkim commented on rabit/rabit/Common/Coordinator/Coordinator.swift:
```diff
@@ -3,8 +3,19 @@ import UIKit
     
     var parentCoordiantor: Coordinator? { get set }
-    var children: [Coordinator] { get }
+    var children: [Coordinator] { get set }
     var navigationController: UINavigationController { get }
     
     func start()
 }
+
+extension Coordinator {
+    func childDidFinish(_ child: Coordinator) {
+        for (index, coordinator) in children.enumerated() {
+            if child === coordinator {
+                children.remove(at: index)
+                break
+            }
+        }
+    }
+}
```
부모 코디네이터로부터 자식 코디네이터를 떼어내는 메소드를 protocol extension으로 구현했습니다. 
모든 코디네이터가 가질 공통적인 로직이고, 프로토콜에 정의되어있는 프로퍼티들만 사용하기 때문에 protocol extension으로 구현하는 방법을 선택했습니다.
그런데 이 메소드에서 `children` 프로퍼티를 수정하는 로직이 필요했고 children이 mutatable 할 필요가 있어 children 프로퍼티의 setter도 제공하도록 했습니다.

Hansolkkim replied:
올리고보니 너무 장풍이라..ㅋㅋㅋ 아래 처럼 변경했습니다!
```swift
extension Coordinator {
    func childDidFinish(_ child: Coordinator) {
        if let childIndex = children.enumerated()
            .first(where: {$1 === child} )?.offset {

            children.remove(at: childIndex)
        }
    }
}
```

junu0516 replied:
ㅋㅋㅋ 확인했어요 ㅎㅎ

junu0516 replied:
이거 배열이 아닌 Set으로 두고, 삽입&삭제를 O(1)로 처리할 수 있을듯 싶어요!

Hansolkkim replied:
`Set<Coordinator>` 를 만드려면 Coordinator가 protocol이 아니라 class 객체여야 될 것 같아요
 Coordinator protocol이 Hashable을 채택해야하는 프로토콜임을 강제해주면 Coordinator 내에 정의되어 있는 parentCoordinator: Coordinator? 프로퍼티에 제약이 생기네요.. 이유를 모르겠어요..ㅠㅠ

junu0516 replied:
아.. 다른 예시 코드들 보면서 왜 O(N)이 되나 했는데 생각해보니 Hashable 채택이 힘들어서 그런거였네요 ㅋㅋ ㅠ 
전에 DrawingApp할 때도 사각형 추상화하면 이거 Hashable한 Key로 사용못해서 어떻게 하나 고민했었는데 같은 문제인 것 같습니다.. 이건 따로 고민좀 해볼께요!

사실 배열에 들어있는 인스턴스 수가 엄청 적어서 ㅋㅋ 시간복잡도는 의미 없는거 같아요 🤣

Hansolkkim commented on rabit/rabit/PhotoEdit/PhotoEditViewModel.swift:
```diff
@@ -76,7 +80,8 @@ private extension PhotoEditViewModel {
 
     func bind(to photoStream: BehaviorRelay<Album.Item>) {
         selectedPhotoData
-            .map { $0 != photoStream.value }
+            .withUnretained(self)
+            .map { ($0.photoEditMode == .addNewPhoto) || ($1 != photoStream.value) }
             .bind(to: saveButtonState)
             .disposed(by: disposeBag)
```
여기서 photoEditMode가 .addNewPhoto의 경우(GoalCoordinator에서 사용하는 경우)에는 무조건 저장버튼이 활성화되어있도록 설정했습니다!

Hansolkkim committed:
chore: Photo의 새로운 초기화생성자 추가

- CertPhotoCameraViewModel에서 새로 찍은 사진의 Photo 모델 객체를 생성할 때 코드를 조금 더 깔끔하게 작성할 수 있고, 초기 디폴트 값을 굳이 CertPhotoCameraViewModel에서 설정해주기 보다는 Photo 생성자에서 넣어주는게 적절하다고 판단하여 새로운 초기화 생성자를 추가

Hansolkkim committed:
chore: Coordinator.childDidFinish(_:) 메소드 수정

Hansolkkim commented on rabit/rabit/Album/Models/Photo.swift:
```diff
@@ -28,6 +28,22 @@ struct Photo: Equatable {
         self.style = style
     }
 
+    // Initializer for newly taken photo
+    init(
+        categoryTitle: String,
+        goalTitle: String,
+        imageData: Data
+    ) {
+        self.init(
+            categoryTitle: categoryTitle,
+            goalTitle: goalTitle,
+            imageData: imageData,
+            date: Date(),
+            color: "#FFFFFF",
+            style: .none
+        )
+    }
+
```
CertPhotoCameraView에서 새로운 사진 정보 저장할 때 초기화 생성자에 불필요하게 여러 값들도 넣어주어야 하는게 보기 안좋아서, 그 상황에서 사용할 수 있는 초기화 생성자 하나 만들었습니다!

junu0516 commented on rabit/rabit/Goal/CertPhotoCamera/CertPhotoCameraViewModel.swift:
```diff
@@ -36,9 +36,9 @@ private extension CertPhotoCameraViewModel {
-            .map { Photo(categoryTitle: goal.category, goalTitle: goal.title, imageData: $0, date: Date(), color: "", style: .none) }
-            .map { BehaviorRelay(value: $0) }
+            .withLatestFrom(previewPhotoData) { (goal.category, goal.title, $1) }
+            .map(Photo.init)
+            .map(BehaviorRelay.init)
```
```suggestion
            .map(BehaviorRelay(value: Photo.init))
```
혹시 가능하면 이렇게 줄이는게 보기 편할 것 같아요!

junu0516 commented on rabit/rabit/Album/Models/Photo.swift:
```diff
@@ -28,6 +28,22 @@ struct Photo: Equatable {
     }
 
+    // Initializer for newly taken photo
+    init(
+        categoryTitle: String,
+        goalTitle: String,
+        imageData: Data
+    ) {
+        self.init(
+            categoryTitle: categoryTitle,
+            goalTitle: goalTitle,
+            imageData: imageData,
+            date: Date(),
+            color: "#FFFFFF",
+            style: .none
+        )
+    }
```
이건 안에서도 생성자 호출하니까 convenience init으로 둬도 괜찮을듯 합니다!

Hansolkkim replied:
저도 처음 알았는데 struct는 convenience initializer를 가지는게 불가능하다고 하네요..ㅠㅠ

junu0516 replied:
헉 그렇군요 ㅋㅋ 저도 몰랐네요

junu0516 commented on rabit/rabit/Goal/CertPhotoCamera/CertPhotoCameraViewModel.swift:
```diff
@@ -36,9 +36,9 @@ private extension CertPhotoCameraViewModel {
-            .withLatestFrom(previewPhotoData)
-            .map { Photo(categoryTitle: goal.category, goalTitle: goal.title, imageData: $0, date: Date(), color: "", style: .none) }
-            .map { BehaviorRelay(value: $0) }
+            .withLatestFrom(previewPhotoData) { (goal.category, goal.title, $1) }
+            .map(Photo.init)
+            .map(BehaviorRelay.init)
```
생성자를 하나 더 추가해주셔서 불필요한 속성을 명시하지 않아도 되서 깔끔하네요 ㅎㅎㅎ
map을 한 번만 쓰도록 개선해도 괜찮을 것 같아요!
(직전에 코드 제안했는데 직접 돌려보니까 컴파일이 안되네요😂 혹시 제가 기존에 map을 2번쓰고 있던거면 제가 고민해볼께요 ㅋㅋㅋㅋ..)

Hansolkkim replied:
네네 원래도 map이 2개이긴 했는데, 이렇게 하는게 가독성이 더 좋을거 같아서 사용했어요ㅎㅎ
Photo 모델을 만든 후 이걸로 BehaviorRelay를 만든다는게 확인되니까 좋은거 같았어요!

junu0516 commented on rabit/rabit/Utilities/UIButton + setBackgroundColor.swift:
```diff
@@ -2,16 +2,14 @@ import UIKit
-              let color = color else { return }
-
-        context.setFillColor(color.cgColor)
-        context.fill(CGRect(x: 0.0, y: 0.0, width: 1.0, height: 1.0))
-
-        let backgroundImage = UIGraphicsGetImageFromCurrentImageContext()
-        UIGraphicsEndImageContext()
+        let backgroundImage = renderer.image { context in
+            color.setFill()
+            context.fill(CGRect(x: 0, y: 0, width: 1.0, height: 1.0))
+        }
 
         self.setBackgroundImage(backgroundImage, for: state)
```
이건 결국 상황에 맞게 비용이 더 저렴한 대안으로 가신걸로 이해해도 될까요??

Hansolkkim replied:
네넵 맞습니다! sRGB 색영역으로 이미지 렌더링시에는 메모리 소모가 크다고 하더라구요!

junu0516 replied:
ㅎㅎㅎ 덕분에 몰랐던 개념 하나 더 알게되었습니다

junu0516 commented on rabit/rabit/Goal/GoalCoordinator.swift:
```diff
@@ -218,51 +185,16 @@ private extension GoalCoordinator {
-        viewController.modalPresentationStyle = .overFullScreen
-        navigationController.present(viewController, animated: true)
-    }
-
-    func dismissStyleSelectView() {
-        navigationController.dismiss(animated: false)
+        photoEditCoordinator.start()
```
195~198번까지는 생성된 Coordinator 구현체를 children 배열에 넣는 것 같은데,
이 부분을 Coordinator 타입을 받는 `addChild(_ coordinator: Coordinator)` 과 같은 메소드로 따로 분기하는건 어떨까요!!?
(개인적으로는 따로 빼서 Coordinator 프로토콜의 extension에 구현부를 정의하면 좋을듯해요 ㅎㅎ)

junu0516 commented on rabit/rabit/PhotoEdit/PhotoEditCoordinator.swift:
```diff
@@ -0,0 +1,180 @@
+import RxRelay
+
+enum PhotoEditMode {
+    case editProperty
+    case addNewPhoto
```
이미 타입명에 Photo, Edit이 둘 다 들어가기 때문에 case는 edit, add 정도로 간단히 둬도 괜찮을 것 같아요!

junu0516 commented on rabit/rabit/PhotoEdit/PhotoEditCoordinator.swift:
```diff
@@ -0,0 +1,180 @@
+
+    private var disposeBag = DisposeBag()
+
+    init(
+        navigationController: UINavigationController,
+        photoStream: BehaviorRelay<Album.Item>,
+        photoEditMode: PhotoEditMode = .editProperty
+    ) {
+        self.navigationController = navigationController
+        self.photoEditMode = photoEditMode
+        self.photoStream = .init(value: photoStream.value)
+        bind(to: photoStream)
+        bind()
+    }
```
본문에서 언급해주신대로 또 다른 코디네이터를 자식으로 두는 방향에 전적으로 동의합니다 ㅎㅎ

다만 한 가지 궁금해지는게 navigationController을 사실상 모든 코디네이터가 생성자 파라미터로 받고 있는데, 개별 코디네이터는 각자의 UINavigationController 인스턴스를 가지거나, 아니면 때에 따라 각자 인스턴스를 가질 지 부모의 인스턴스를 물려받을 지 한 가지 기준을 선택해야 되지 않을까 싶어요!
(사실상 후자의 방향으로 자연스레 흘러온 것 같은데, 만약에 구조에 제약을 좀 더 엄격히 둘 거면 한 번 UINavigationController 소유 및 생성 시점을 정리하고 가면 좋을 것 같습니다 ㅎㅎ 아무래도 코디네이터 구현체들마다 생성자가 점점 상이해지는 것 같아서요!)

현재 프로젝트 구조상

- SceneDelegate에서 UINavigationController 인스턴스 생성
- AppCoordinator은 SceneDelegate에서 인스턴스를 생성자 파라미터로 받아서 속성 초기화
- MainCoordinator은 AppCoordinator이 가진 navigationController 인스턴스를 마찬가지로 생성자 파라미터로 받아서 속성 초기화
- 다시 개별 텝에 해당하는 Goal, AlbumCoordinator은 각자의 UINavigationController 인스턴스를 생성해서 속성 초기화

이런식으로 navigationController 속성을 초기화하고 있는데, 앞서 말한대로 상황에 따라 부모로 부터 물려받거나 직접 자신의 인스턴스를 생성할 지 선택할 지, 혹은 인스턴스 생성에 대한 제약을 둘 지 이야기해보고 싶습니다 ㅎㅎ

Hansolkkim replied:
좋은 생각인거 같아요!
그런데 PhotoEditCoordinator 같은 경우에, 부모로부터 navigationController를 물려받았지만 사용은 안하고 있는데
Coordinator에서 새로운 VC를 띄우더라도 결국은 `navigation.present(_:animated:)` 메소드를 사용해 띄워야하지 않을까 싶은데, 이런 경우에도 이전 Coordinator에서 띄운 navigation을 알아야 present 할 수 있었습니다!

push/pop을 이용하지 않는 Coordinator에서도 일단은 띄우려는 화면을 띄우기 위해서는 현재 띄워져 있는 화면이 무엇인지(화면을 띄우려는 주체가 누구인지) 알아야하는 과정이 필요해서 navigationController를 물려주는 초기화 생성자가 필요했어요!

junu0516 replied:
아하 이해했습니당 ㅎㅎ
주말지나고 나서 생각해보니 구조에 제약을 두는게 오히려 더 힘들고 비용이 더 클 것 같아서... UINavigationController의 생성 및 소유는 코디네이터별로 상황에 맞게 유동적으로 처리하면 될듯하네요 !! ㅋㅋ

junu0516 replied:
현재 코디네이터가 Coordinator 프로토콜로 추상화된 상태라서 혹시나 자식 코디네이터 붙이는 작업도 (저희 전에 뷰컨트롤러 띄우는거 추상화 얘기해봤듯이) 제네릭 + 프로토콜 조합 사용하는 식으로 추상화할 경우에는, photoStream, photoEditMode 같은 부분을 어떻게 전달할 지 고민해봐야 하나 싶어서 코멘트 달았었어요!!

다만.. 이게 저번처럼 생각보다 적용하는 데 비용이 더 크지 않을까 싶기도 해서(화면이 그닥 많지도 않고) 지금처럼 사용하는게 더 나을 것 같습니다 ㅎㅎ

junu0516 reviewed (approved):
수고 많으셨어요 ㅎㅎ
생각나는 지점이랑 궁금한 점들을 코멘트로 달았습니다 ㅎㅎ!

코드를 좀 실제로 돌려보고 싶어서.. 일단 코멘트로 서로 커뮤니케이션 하면서 추가로 궁금하거나 생각나는거 있음 또 물어볼께요!

Hansolkkim committed:
chore: [#99] PR 리뷰 반영

- Coordinator protocol extension으로 `addChild(_:)` 메소드 추가하여, 모든 Coordinator에서 공통적으로 사용할 Child Coordinator를 추가하는 메소드 구현
- PhotoEditMode의 케이스 네이밍 수정 (editProperty -> edit / addNewPhoto -> add)

Hansolkkim commented:
@junu0516 
리뷰 반영 바로 했습니다ㅎㅎ

AlbumCoordinator와 PhotoEditCoordinator 관련해서는 고민을 좀 해봤는데,
Delegate 패턴을 이용해서 `PhotoEditCoordinatorDelegate`와 같은 프로토콜을 선언하고 AlbumCoordinator가 이를 채택하도록 하여 하위 코디네이터인 PhotoEditCoordinator가 특정 상황에 상위 코디네이터인 AlbumCoordinator: PhotoEditCoordinatorDelegate의 특정 메소드를 호출하도록 하든지, 아니면 이전처럼 특정 프로퍼티를 호출하는 것도 방법이 되겠다 생각이 들었습니다ㅎㅎ

junu0516 commented:
> @junu0516 리뷰 반영 바로 했습니다ㅎㅎ
> 
> AlbumCoordinator와 PhotoEditCoordinator 관련해서는 고민을 좀 해봤는데, Delegate 패턴을 이용해서 `PhotoEditCoordinatorDelegate`와 같은 프로토콜을 선언하고 AlbumCoordinator가 이를 채택하도록 하여 하위 코디네이터인 PhotoEditCoordinator가 특정 상황에 상위 코디네이터인 AlbumCoordinator: PhotoEditCoordinatorDelegate의 특정 메소드를 호출하도록 하든지, 아니면 이전처럼 특정 프로퍼티를 호출하는 것도 방법이 되겠다 생각이 들었습니다ㅎㅎ

@Hansolkkim 
요건 혹시 하위가 상위에 동작 끝난거 알릴 때 Rx대신 Delegate 패턴으로 대체하는 걸로 이해하면 될까요!!?
Delegate 패턴 쓰면 ㅋㅋ 제안하셨던 네이밍이 훨씬 자연스러울 것 같아요!!

하나 궁금한게 보통 RxSwift 사용하는 프로젝트에서 Rx 바인딩이랑 Delegate 패턴 중에 어떤 식으로 객체 간에 서로 소통하는 방식이 보편적인가 하는 생각이 드네요ㅎㅎㅎ
(이전에 이슈트래커 할 때 Deleagte -> Rx로 수정했던 기억이 있는데 지금 상황은 제가 맞게 이해한거면 그 때랑 반대길래 보통 현업에서 어떻게 할까 생각들었습니다 ㅎㅎ)

Hansolkkim commented:
@junu0516 

음 아직 대체하기러 한건 아닌데, Rx를 사용해서 하려면 Protocol을 하나 더 만들어야 하고, 그 프로토콜의 역할이 Navigation과 동일하게 Delegate 패턴의 느낌이라 그런식으로 하는게 제일 자연스럽지 않을까 생각했어요ㅋㅋ
사실 Rx 바인딩을 사용했을 뿐이지 똑같이 Delegate 패턴이라 말씀은 그렇게 드렸습니다ㅎㅎ
(Navigation Protocol들과 명칭상 차이를 두기 위해서 `~~Delegate` 이라는 프로토콜 명을 생각했었던거 같아요!)

junu0516 commented:
> @junu0516
> 
> 음 아직 대체하기러 한건 아닌데, Rx를 사용해서 하려면 Protocol을 하나 더 만들어야 하고, 그 프로토콜의 역할이 Navigation과 동일하게 Delegate 패턴의 느낌이라 그런식으로 하는게 제일 자연스럽지 않을까 생각했어요ㅋㅋ 사실 Rx 바인딩을 사용했을 뿐이지 똑같이 Delegate 패턴이라 말씀은 그렇게 드렸습니다ㅎㅎ (Navigation Protocol들과 명칭상 차이를 두기 위해서 `~~Delegate` 이라는 프로토콜 명을 생각했었던거 같아요!)

넵ㅋㅋ 이해했습니다!
코드 쭉 돌려봤고 수정하신 커밋도 확인했어요!
더 이상 수정하거나 추가하실거 없으시면 머지하셔도 될듯합니다~!
(나머지 구조 관련해서는 추후에 직접 리팩토링 들어갈 때 더 얘기해면 될듯해요)
